### PR TITLE
Done below points:

### PR DESCRIPTION
**Pull request by KotiGungi-Vienna**
1.Add Currency Rate Type Field on the Bank Statement Line and this field will be UI Mandatory.  

2.On the Statement Line, when payment Id reference is present then currency rate type should be same as on the Payment window. 

3.On the Statement Line, when Payment ID Reference is present then Currency Rate type field will be read only. 

4.On the Statement Line when we select Payment then the system will set same account date as on the Payment window.  

          Note: At the time of changing statement date it should Change the Account date. 

5.On the Statement Line when Invoice ID reference is present then Payment will generate with Currency Rate type bind on the line. 

6.On the Statement Line when Invoice ID Reference is present then payment will generate with Invoice Currency. 

 7.On the Statement Line when Order ID reference is present then Payment will generate with Currency Rate type bind on the Line. 

8.On the Statement Line when Order ID reference is present then payment will generate with Order Currency.

### CODE REVIEW

File: ModelLibrary/Model/MBankStatementLine.cs [hunk 1]
```diff
@@ -113,6 +113,70 @@ public void SetPayment(MPayment payment)
...
+                {
+                    _InvCount = Util.GetValueOfInt(DB.ExecuteScalar("SELECT COUNT(C_PaymentAllocate_ID) FROM C_PaymentAllocate WHERE IsActive='Y' AND C_Payment_ID=" + payment.GetC_Payment_ID(), null, null));
+                }
+                else 
+                {
+                    _InvCount = payment.GetC_Invoice_ID();
+                }
+                if (_InvCount > 0 || payment.GetC_Order_ID() > 0)
+                {
+                    SetVA012_VoucherType("M");
```
Comment (Amit-Vienna) [inline ModelLibrary/Model/MBankStatementLine.cs:170]: M belongs to ?

File: ModelLibrary/Process/BankStatementPayment.cs [hunk 2]
```diff
@@ -96,7 +97,7 @@ private String CreatePayment(X_I_BankStatement ibs)
             MPayment payment = CreatePayment(ibs.GetC_Invoice_ID(), ibs.GetC_BPartner_ID(),
                 ibs.GetC_Currency_ID(), ibs.GetStmtAmt(), ibs.GetTrxAmt(),
                 ibs.GetC_BankAccount_ID(),Utility.Util.GetValueOfDateTime(ibs.GetStatementLineDate() == null ? ibs.GetStatementDate() : ibs.GetStatementLineDate()),
-                Utility.Util.GetValueOfDateTime(ibs.GetDateAcct()), ibs.GetDescription(), ibs.GetAD_Org_ID());
+                Utility.Util.GetValueOfDateTime(ibs.GetDateAcct()), ibs.GetDescription(), ibs.GetAD_Org_ID(), 0, 0); //Used Zero's as parameters to Avoid throw Error
             if (payment == null)
             {
                 throw new SystemException("Could not create Payment");
```
Comment (Amit-Vienna) [inline ModelLibrary/Process/BankStatementPayment.cs:103]: need actual message, why payment not created

File: ModelLibrary/Process/BankStatementPayment.cs [hunk 3]
```diff
@@ -213,53 +249,223 @@ private String CreatePayment(MBankStatementLine bsl)
...
+                                    {
+                                        error = pp != null ? pp.ToString() : "";
+                                    }
+                                }
+                                _message = !string.IsNullOrEmpty(error) ? error : "VA012_PaymentNotSaved";
+                                return null;
+                            }
+                            else
+                            {
+                                payment.SetPayAmt(payment.GetPayAmt() + PayAlocate.GetAmount());
```
Comment (Amit-Vienna) [inline ModelLibrary/Process/BankStatementPayment.cs:332]: this else section not required, system will do it on after save functionality

File: ModelLibrary/Process/BankStatementPayment.cs [hunk 4]
```diff
@@ -213,53 +249,223 @@ private String CreatePayment(MBankStatementLine bsl)
...
+                payment.SetC_Currency_ID(invoice.GetC_Currency_ID());
+                decimal _dueAmt = 0;
+
+                string _sql = "SELECT * FROM C_InvoicePaySchedule WHERE VA009_PAIDAMNT IS NULL AND VA009_IsPaid = 'N' AND IsActive = 'Y' AND C_Invoice_ID =" + invoice.GetC_Invoice_ID();
+                DataSet _ds = DB.ExecuteDataset(_sql, null, Get_Trx());
+
+                if (_ds != null && _ds.Tables[0].Rows.Count == 1)
                 {
-                    payment.SetC_Currency_ID(C_Currency_ID);
+                    payment.SetC_InvoicePaySchedule_ID(Util.GetValueOfInt(_ds.Tables[0].Rows[0]["C_INVOICEPAYSCHEDULE_ID"]));
```
Comment (Amit-Vienna) [inline ModelLibrary/Process/BankStatementPayment.cs:267]: from where we set the invoice reference

File: ModelLibrary/Process/BankStatementPayment.cs [hunk 5]
```diff
@@ -193,6 +227,8 @@ private String CreatePayment(MBankStatementLine bsl)
             payment.SetAD_Org_ID(AD_Org_ID);
             payment.SetC_BankAccount_ID(C_BankAccount_ID);
             payment.SetTenderType(MPayment.TENDERTYPE_Check);
```
Comment (Amit-Vienna) [inline ModelLibrary/Process/BankStatementPayment.cs:229]: why we hard code tender type ?

File: ModelLibrary/Process/BankStatementPayment.cs [hunk 6]
```diff
@@ -213,53 +249,223 @@ private String CreatePayment(MBankStatementLine bsl)
...
+                    {
+                        Get_Trx().Rollback();
+                        return null;
+                    }
+                    else
+                    {
+                        for (int i = 0; i < _ds.Tables[0].Rows.Count; i++)
+                        {
+
+                            MPaymentAllocate PayAlocate = new MPaymentAllocate(GetCtx(), 0, Get_Trx());
```
Comment (Amit-Vienna) [inline ModelLibrary/Process/BankStatementPayment.cs:296]: initialize object outside the loop

File: ModelLibrary/Process/BankStatementPayment.cs [hunk 7]
```diff
@@ -213,53 +249,223 @@ private String CreatePayment(MBankStatementLine bsl)
...
+                            PayAlocate.SetC_Payment_ID(payment.GetC_Payment_ID());
+                            PayAlocate.SetAD_Client_ID(payment.GetAD_Client_ID());
+                            //set Organization with the reference of Bank Account
+                            PayAlocate.SetAD_Org_ID(invoice.GetAD_Org_ID());
```
Comment (Amit-Vienna) [inline ModelLibrary/Process/BankStatementPayment.cs:300]: update org reference based on payment header org

**Review by Amit-Vienna — CHANGES_REQUESTED** [review]
do the needful as per comments

**Comment by KotiGungi-Vienna** [comment]
Okay sir!


Best Regards
Koteswararao






From: Amit Vienna ***@***.***>
To: VIENNA-Advantage-ERP-CRM/official-VAFramework-ERP-CRM 
***@***.***>
Cc: KotiGungi-Vienna ***@***.***>,  Author 
***@***.***>
Date: Thu, 01 Apr 2021 00:09:02 -0700
Subject: Re: [VIENNA-Advantage-ERP-CRM/official-VAFramework-ERP-CRM] Done 
below points: (#876)



CAUTION: This email originated from outside of the organization. Do not 
click links or open attachments unless you recognize the sender and know the 
content is safe.







@Amit-Vienna requested changes on this pull request.

do the needful as per comments

File: ModelLibrary/Process/BankStatementPayment.cs (hunk 2 above)
Comment (KotiGungi-Vienna) [reply ModelLibrary/Process/BankStatementPayment.cs:103]: Here not need Change anything sir, we are not using this Method as you said, for present requirement using CreatePayment(MBankStatementLine bsl) not CreatePayment(X_I_BankStatement ibs)!

File: ModelLibrary/Model/MBankStatementLine.cs (hunk 1 above)
Comment (KotiGungi-Vienna) [reply ModelLibrary/Model/MBankStatementLine.cs]: done sir!

File: ModelLibrary/Process/BankStatementPayment.cs (hunk 3 above)
Comment (KotiGungi-Vienna) [reply ModelLibrary/Process/BankStatementPayment.cs]: removed else section!

File: ModelLibrary/Process/BankStatementPayment.cs (hunk 4 above)
Comment (KotiGungi-Vienna) [reply ModelLibrary/Process/BankStatementPayment.cs:270]: done!

File: ModelLibrary/Process/BankStatementPayment.cs (hunk 5 above)
Comment (KotiGungi-Vienna) [reply ModelLibrary/Process/BankStatementPayment.cs]: Done!

File: ModelLibrary/Process/BankStatementPayment.cs (hunk 6 above)
Comment (KotiGungi-Vienna) [reply ModelLibrary/Process/BankStatementPayment.cs]: Done!

File: ModelLibrary/Process/BankStatementPayment.cs (hunk 7 above)
Comment (KotiGungi-Vienna) [reply ModelLibrary/Process/BankStatementPayment.cs]: Done!

**Review by Amit-Vienna — APPROVED** [review]
fine

**Pull request merged**